### DISCUSSION TITLE
Code question: while loop inside the default progress callback lambda

### DISCUSSION THREAD

**Discussion by crasm** [ggml-org/llama.cpp#2864] [General]
[llama.cpp:L5370](https://github.com/ggerganov/llama.cpp/blob/44c117f41ee01c5ac8fb86bba041f08d8b87b46d/llama.cpp#L5370)

Why is it a while loop instead of an if statement? Is this for a compiler optimization or something? Changing it to an if statement seems to have the exact same printing behavior (casually testing with `main --no-mmap`). I don't see how the loop would execute more than once, given that cur_percentage_p is updated within the loop immediately after the check.

(I am attempting to write dart bindings for llama.cpp. I can't tell if this is hastily written code or if it has some subtle behavior that is going over my head.)

```
    unsigned cur_percentage = 0;
    if (params.progress_callback == NULL) {
        params.progress_callback_user_data = &cur_percentage;
        params.progress_callback = [](float progress, void * ctx) {
            unsigned * cur_percentage_p = (unsigned *) ctx;
            unsigned percentage = (unsigned) (100 * progress);
            while (percentage > *cur_percentage_p) {
                *cur_percentage_p = percentage;
                LLAMA_LOG_INFO(".");
                if (percentage >= 100) {
                    LLAMA_LOG_INFO("\n");
                }
            }
        };
    }
 ```

**Comment by ggerganov** [comment]
It's hastily written. Technically `*cur_percentage_p = percentage;` should be something like `*cur_percentage_p += step;` so that we print the same number of `"."` for different model sizes, but it's just a display thing so not really important. The way it is, `if` is better than `while`

**Comment by crasm** [comment]
Ah, maybe I'll take a look at implementing that since I'm digging around there. Thank you for clarifying
